test/README.md

```diff
@@ -42,26 +42,42 @@ The test suite contains these kind of files:
    ```
    to allow any number of type definitions followed by encoding tests with grammar
    ```
-   <test> = <argseq> <input> <outcome> <desc>
-   <argseq>  ::= ( <argtype>,* )
-   <input> =  <text> | blob <text>
-   <outcome> = true | false
-   <desc> = <text>?
+   <test> ::=
+      | assert <input> : <tuptype> <desc>
+      | assert <input> !: <tuptype> <desc>
+      | assert <input> == <input> : <tuptype> <desc>
+      | assert <input> != <input> : <tuptype> <desc>
+   <input> ::= <text> | blob <text>
```

Comment:
```suggestion
   <input> ::= <val> | blob <text>
```

Reply:
I thought about this, but I have doubts: Consider a tool that code-generates a test suite. This tool parses the `.test.did` file, and then generate host language code. So we have a two-stage process (this is actually the case in `haskell-candid` as of today; it uses TH to parse the `foo.test.did` file at compile-time, and generates assertions at the right Haskell type.)

It is clear that the parsing of Candid binary values needs to happen in the second stage – after all, the test suite should be generated even if some binaries values are invalid (e.g. "DADL\00\00").

But it seems we should treat the textual notation here analogously: Parsing the the assertions should _not_ require parsing the test _input_, which is data.

It is also a prerequistie to do thing like
```
assert "(true" !: (bool) "missing parenthesis"
```
or
```
assert "{actor = null}" !: (record { "actor" : null}) "unescaped keyword"
```

Reply:
Interesting, so the `!=`  means either not equal or syntax error. Maybe we need a `assert_parse` for the later? Usually, `assert` assumes the inputs are syntactically correct, e.g. it's fine to test `blob "DADL\00\00"`, but not `bob "DADL`.

Reply:
> Maybe we need a assert_parse for the later?

It seems like it's less about _positively asserting_ the need for a correct parse (it's _always_ needed implicitly for all tests that include text values, so not useful info to assert this need explicitly).

Rather, I think we want to indicate when we do not expect a parse to succeed (`assert_bad_parse`, or something?).  That does seem useful, and missing here.

Reply:
> It is also a prerequistie to do thing like [...]

Those examples would dictate official error messages, as strings, for each parse error?

If that's the intention, it seems like:
 - you don't really need the type in those cases, and 
 - you may want to indicate the intention of the test in the assertion kind (e.g., `assert_parse_error`, or something)

Reply:
Oh, nevermind --- I think those test case descriptions are not meant to be interpreted by any tools, just humans.  (Right?)

Reply:
Allowing plain `val` would make sense as a convenience short-hand, but not instead of, but in addition to the string form -- the latter is still needed for negative tests, as @nomeata says.

@chenyan-dfinity, `:` and `!:` _are_ assert_parse and assert_no_parse, respectively.

Reply:
> Interesting, so the != means either not equal or syntax error. Maybe we need a assert_parse for the later? Usually, assert assumes the inputs are syntactically correct, e.g. it's fine to test blob "DADL\00\00", but not bob "DADL.

No, `!=` means: Both parse, but to different values. `!:` means assert_no_parse (but does not and should not  distiguish betwen “is totally bogus” and “might parse at a different type”).

> I think those test case descriptions are not meant to be interpreted by any tools, just humans. (Right?)

Yes, see the example output of a test suite futher up for how the description is useful.